presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -361,14 +361,25 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
                 return process(coerceDoubleToLongComparison(normalized), complement);
             }
 
-            if (!TypeRegistry.canCoerce(value.getType(), fieldType)) {
+            Optional<NullableValue> coercedValue = coerce(value, fieldType);
+            if (!coercedValue.isPresent()) {
                 return super.visitComparisonExpression(node, complement);
             }
 
-            Object coerced = new FunctionInvoker(metadata.getFunctionRegistry())
-                    .invoke(metadata.getFunctionRegistry().getCoercion(value.getType(), fieldType), session.toConnectorSession(), value.getValue());
+            return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+        }
 
-            return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coerced, complement);
+        private Optional<NullableValue> coerce(NullableValue value, Type targetType)
+        {
+            if (!TypeRegistry.canCoerce(value.getType(), targetType)) {
+                return Optional.empty();
+            }
+            if (value.isNull()) {
+                return Optional.of(NullableValue.asNull(targetType));
+            }
+            Object coercedValue = new FunctionInvoker(metadata.getFunctionRegistry())
```

Comment:
We should make FunctionInvoker aware of `@nullable` argument annotations and short circuit if they are not present (per standard sql semantics)